implementations/micrometer-registry-graphite/src/test/java/io/micrometer/graphite/GraphiteTagNameMapperTest.java

```diff
@@ -0,0 +1,52 @@
+/**
+ * Copyright 2017 Pivotal Software, Inc.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.graphite;
+
+import io.micrometer.core.Issue;
+import io.micrometer.core.instrument.Meter;
+import io.micrometer.core.instrument.config.NamingConvention;
+import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Tests for {@link GraphiteTagNameMapper}.
+ *
+ * @author Jon Schneider
+ * @author Johnny Lim
+ * @author Andrew Fitzgerald
+ */
+class GraphiteTagNameMapperTest {
```

Comment:
TODO: Add tests/logic for valid chars for graphite tags/values
https://graphite.readthedocs.io/en/latest/tags.html#carbon

>Tag names must have a length >= 1 and may contain any ascii characters except `;!^=`. Tag values must also have a length >= 1, and they may contain any ascii characters except `;~`.

Reply:
Will we have to update `GraphiteNamingConvention` to put a placeholder tag value like `unknown` in when the tag value is blank?

Reply:
>Will we have to update GraphiteNamingConvention to put a placeholder tag value like `unknown` in when the tag value is blank?

I believe so (although I think i'm leaning towards `unspecified`).


>Also, should update the naming convention to bar the use of ; and =.

👍

Reply:
So I added a (failing) test suite in 612566a1ddf4740d8a6485083064740873e1c88c.

After we dropped the prefix tags, I think it makes sense to use a new naming convention for tag based Graphite metrics.

I don't think camel casing makes since any more, instead we should honor the metric name passed in by the user unless there are invalid characters.

Reply:
I'm thinking I did that wrong, and those tests should really be for a new `GraphiteHierarchicalNamingConvention`

Reply:
> I don't think camel casing makes since any more, instead we should honor the metric name passed in by the user unless there are invalid characters.

Agreed, the only reason for camel casing was so they could do wildcard queries for hierarchical names where the tags embedded in the name were separating by dots. So the dot syntax is fine now with the richer tags.